dev/io.openliberty.security.jakartasec.4.0.internal/src/io/openliberty/security/jakartasec40/identitystore/InMemoryIdentityStore.java

```diff
@@ -0,0 +1,161 @@
+/*******************************************************************************
+ * Copyright (c) 2025 IBM Corporation and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License 2.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+package io.openliberty.security.jakartasec40.identitystore;
+
+import java.util.HashSet;
+import java.util.Set;
+
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.ConfigurationPolicy;
+
+import com.ibm.websphere.ras.ProtectedString;
+import com.ibm.websphere.ras.Tr;
+import com.ibm.websphere.ras.TraceComponent;
+import com.ibm.websphere.ras.annotation.Sensitive;
+
+import io.openliberty.security.jakartasec40.identitystore.InMemoryIdentityStoreDefinitionWrapper.CredentialValue;
+import jakarta.enterprise.context.ApplicationScoped;
+import jakarta.enterprise.inject.Default;
+import jakarta.security.enterprise.credential.CallerOnlyCredential;
+import jakarta.security.enterprise.credential.Credential;
+import jakarta.security.enterprise.credential.UsernamePasswordCredential;
+import jakarta.security.enterprise.identitystore.CredentialValidationResult;
+import jakarta.security.enterprise.identitystore.IdentityStore;
+import jakarta.security.enterprise.identitystore.InMemoryIdentityStoreDefinition;
+
+/**
+ * Liberty's InMemory {@link IdentityStore} implementation.
+ */
+@Default
+@ApplicationScoped
+@Component(service = {},
+           immediate = true,
+           configurationPolicy = ConfigurationPolicy.IGNORE,
+           property = "service.vendor=IBM")
+public class InMemoryIdentityStore implements IdentityStore {
+
+    private static final TraceComponent tc = Tr.register(InMemoryIdentityStore.class);
+
+    /** The definitions for this IdentityStore. */
+    private final InMemoryIdentityStoreDefinitionWrapper inMemoryIdentityStoreDefinitionWrapper;
+
+    /**
+     * Construct a new {@link InMemoryIdentityStore} instance using the specified definitions
+     * from the annotation.
+     *
+     * @param inMemoryIdentityStoreDefinition The definitions to use to configure the {@link IdentityStore}.
+     */
+
+    public InMemoryIdentityStore(InMemoryIdentityStoreDefinition inMemoryIdentityStoreDefinition) {
+        inMemoryIdentityStoreDefinitionWrapper = new InMemoryIdentityStoreDefinitionWrapper(inMemoryIdentityStoreDefinition);
+    }
+
+    @Override
+    public Set<String> getCallerGroups(CredentialValidationResult validationResult) {
+        Set<String> groups = new HashSet<String>();
+        if (!validationTypes().contains(ValidationType.PROVIDE_GROUPS)) {
+            return groups; // empty
+        }
+
+        String caller = validationResult.getCallerPrincipal().getName();
```

Comment:
Can getCallerPrincipal() return null?  NullPointerException would happen in that case.